pkg/cache/cluster.go

```diff
@@ -403,63 +398,77 @@ func (c *clusterCache) watchEvents(ctx context.Context, api kube.APIResourceInfo
 			if err != nil {
 				return fmt.Errorf("failed to load initial state of resource %s: %v", api.GroupKind.String(), err)
 			}
-			c.replaceResourceCache(api.GroupKind, info.resourceVersion, items, ns)
-			return nil
-		})
 
-		if err != nil {
-			return err
+			err = runSynced(&c.lock, func() error {
+				c.replaceResourceCache(api.GroupKind, items, ns)
+				return nil
+			})
+
+			if err != nil {
+				return err
+			}
 		}
 
-		w, err := resClient.Watch(ctx, metav1.ListOptions{ResourceVersion: info.resourceVersion})
+		w, err := watchutil.NewRetryWatcher(resourceVersion, &cache.ListWatch{
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return resClient.Watch(ctx, options)
+			},
+		})
+
 		if errors.IsNotFound(err) {
```

Comment:
If you dig into `watchutil.NewRetryWatcher()` you will see there is no way it can return a NotFound error. In fact, it contacts the remote side asynchronously. I've tried to do the same refactoring as I believe it's very necessary, but this is the blocker I've ran into. I think the case where the watched API went away is definitely a possibility, but not as common. Perhaps there should be some out of bound process that resynchronizes the list of APIs and stops the no longer needed watches (and starts the needed ones, maybe)?

Reply:
I still think this is an improvement and should be merged :)

Reply:
Thank you! We can catch that error in `WatchFunc` implementation. Done.